org.eclipse.jface.text/src/org/eclipse/jface/text/contentassist/AsyncCompletionProposalPopup.java

```diff
@@ -427,7 +433,11 @@ public void run() {
 							fMutex.wait(fAutoActivationDelay);
 					}
 					Optional<Display> display= Optional.ofNullable(fContentAssistSubjectControlAdapter.getControl()).map(Control::getDisplay);
-					display.ifPresent(d -> d.asyncExec(() -> displayProposals(true)));
+					display.ifPresent(d -> d.asyncExec(() -> {
+						if (hasProposals()) {
+							displayProposals(true);
```

Comment:
Do I get it right that this wouldn't show the popup while proposals are computing? IMO, it's interesting to show the "Computing" feedback to users.
I think the condition should be `if (stillComputing() || hasProposls())` and then, if there are no proposals `if (!stillComputing() && !hasProposals() && !isAutoActivated())` then hide the popup.

Reply:
Yes, I agree that if the computing has been requested, the feedback can be shown to the user. I will update the PR to avoid the popup only if is autoactivated.

Reply:
@mickaelistria , I have checked the code again. The condition `!isAutoActivated()` is implicit in the `run` method, because the `PopupVisibleTimer` is only started if `!autoActivated`, this is show in the excerpt:

```

		displayProposals(!autoActivated);
	}

	@Override
	void displayProposals(boolean showPopup) {
		if (showPopup) {
			fPopupVisibleTimer.stop();
		}

		super.displayProposals(showPopup);
		if (!showPopup) {
			fPopupVisibleTimer.start();
		}
	}
```

So, I think the PR is fine from this point of view. I will write a second comment on the other review item: if the computing has been requested, the feedback can be shown to the user. I knew that when I coded the PR, but forgot over the weekend about it :)

Reply:
> Do I get it right that this wouldn't show the popup while proposals are computing? IMO, it's interesting to show the "Computing" feedback to users. I think the condition should be `if (stillComputing() || hasProposls())` and then, if there are no proposals `if (!stillComputing() && !hasProposals() && !isAutoActivated())` then hide the popup.

After looking at this again, I think one can actually remove the `PopupVisibleTimer` that was introduced in https://github.com/eclipse-platform/eclipse.platform.text/commit/78fc9f6f6566323c691f1fee1eb14de2f27e7d34. By the comment on that commit:

> if async completion is triggered automatically, then the popup is only shown if there are completions to be shown

but that is not what happens if the `PopupVisibleTimer`, the timer as introduced in that commit, will just show the popup after the `fAutoActivationDelay` of 500 passes. The code in the lambda `computeAndPopulateProposals` for the async case will already show the popup the first time there is a completion proposal added, and it will also close it if it is open and the last answer arrives.

My PR adds the same condition in the timer as in the `computeAndPopulateProposals`, so I think it is correct as it is, but we could as well remove the timer, or am I overlooking something. @gayanper , @vogella , you worked on the commit that introduced the problem. Do you have any opinion on my comment that we could remove the timer or if we should apply the condition as in this PR or as suggested by @mickaelistria?

Thanks